[FLOC-3829] Have the GCE driver implement ICloudAPI

sarum90 opened the pull request:
Implement the ICloudAPI in the GCE IBlockDeviceAPI.

Also, actually test all of the endpoints at the functional level.

The GCE functional tests now are considerably slower to run since they actually spin up 2nd instances for some tests.  Since they are only run nightly on Jenkins this seemed reasonable.

justnoise commented on flocker/node/agents/gce.py:
```diff
@@ -350,7 +353,7 @@ def _disk_resource_description(self):
         return u"flocker-v1-cluster-id: " + unicode(self._cluster_id)
 
-    def _do_blocking_operation(self, function, **kwargs):
+    def _do_blocking_operation(self, function, max_wait=60, **kwargs):
```
:grey_question: Nitpick: timeout is a more typical way of expressing max_wait.

Yes, I had to find _something_ in here :)

sarum90 replied:
Lol, Thanks-fixed. Actually that is definitely something I would have complained about later. I sincerely appreciate you catching that before I put that into master.

justnoise commented:
I can't find much to fault in here.  Implementation seems pretty close to what's done for EBS and Cinder with the addition of walking GCE response pages.

wallrj commented on flocker/node/agents/functional/test_gce.py:
```diff
@@ -78,7 +81,18 @@ def create_instance(self, instance_name):
+                    # The test must have already destroyed the instance.
+                    pass
+                else:
+                    raise e
```
You probably know this already, but you lose the stack trace if you `raise e`. Instead just `raise`.
 * http://www.ianbicking.org/blog/2007/09/re-raising-exceptions.html

And maybe do 

```
if e.resp.status != 404:
    raise
```

...while you're here.

sarum90 replied:
Thanks you for the reminder! When I wrote that I felt like I was doing it wrong.

wallrj commented on flocker/node/agents/functional/test_gce.py:
```diff
@@ -78,7 +81,18 @@ def create_instance(self, instance_name):
+                else:
+                    raise e
+
+        self.addCleanup(lambda: destroy_best_effort(instance))
```
:question: nit this could be `self.addCleanup(destroy_best_effort, instance)` I think

sarum90 replied:
+1, But somehow I missed this comment before I kicked off my most recent test branch.

Is there a technical reason to prefer one over the other?

wallrj commented on flocker/node/agents/gce.py:
```diff
@@ -544,3 +549,42 @@ def destroy_volume(self, blockdevice_id):
+            nodes.extend(result.get('items', []))
+            done = not page_token
+        return set(node.get("name") for node in nodes
+                   if node.get("status") == "RUNNING")
```
:question: Might be better to use `node["name"]` and `node["status"]` so we get an clear error if the keys of the node dictionary ever change.

sarum90 replied:
Done (above as well)

wallrj commented on flocker/node/agents/gce.py:
```diff
@@ -544,3 +549,42 @@ def destroy_volume(self, blockdevice_id):
+            self._compute.instances().stop,
+            timeout_sec=5*60,
+            instance=node_id
+        )
```
:question: Sounds like a useful method that should eventually be part of the ICloudAPI if we want to share the test_list_live_nodes test across other drivers. 
For now, I wonder if it should be private so that it's clearly not part of ICloudAPI.

sarum90 replied:
+1 I also winder if this should actually be part of the provisioner instead of part of this interface.

For now though, just doing a GCE-specific implementation.

Changing to private though, as it is only used by this and the immediate functional tests of this.

sarum90 replied:
Also start_node and what is now _stop_node are only ever used in tests.  Really `ICloudAPI` is known to be an ugly interface at this point.  Somewhere that is cited for the reason that start_node isn't properly tested in the make_icloudapi_tests test suite :/

wallrj commented on flocker/node/agents/functional/test_gce.py:
```diff
@@ -144,6 +169,64 @@ def test_multiple_cluster(self):
+        """
+        list_live_nodes should be able to walk pages to get all live nodes and
+        should not have nodes after they are destroyed.
+        """
```
:question: Why bother testing pagination here? I'd remove that part.
Seems like this'd be a useful `ICloudAPI` test in the future.
Other drivers probably don't / can't support the `stop_node` API so perhaps it should just rely on creating and terminating test nodes.

sarum90 replied:
So spinning up a second node is expensive, and I need at least 2 live nodes in order to verify pagination is working... sort of bad, but I went for 1 large large test that tests everything that requires having a second node spun up rather than two proper unit tests.

I will ~~make a sacrifice to the testing gods in order to ask for forgiveness for this uncleanliness~~ add a comment reflecting this.

wallrj commented:
Thanks @sarum90 and @justnoise 

I left a few comments, some optional, some questions.

Please answer or address those and merge when green.

I also left a comment about the `already_attached` tests situation https://clusterhq.atlassian.net/browse/FLOC-1839

wallrj commented:
Oh, and I forgot to say, do you think it'd be worth making the GCE volume description a JSON blob from the get go to allow extra information to be added in future? 
Here:
```
    def _disk_resource_description(self):
        """
        Returns the value to be used in the description field of the disk
        resources for this cluster.

        :returns unicode: The value for the description.
        """
        return u"flocker-v1-cluster-id: " + unicode(self._cluster_id)

```

sarum90 commented:
Thanks for the review! I addressed comments, and plan on re-kicking the test on jenkins, waiting for green and merging.

Re the GCE volume description:

GCE filters on those fields are really basic, essentially only allowing for exact string match. Thus, I'd like the content of description to be a pure function of the cluster-id in case we need to filter by cluster id. The driver does not currently use GCE filters, but for now I did not want to put us in a position where we would be unable to use filter on cluster id in the future (This would make getting a list of all volumes in a given cluster faster). I have a feeling if it were a JSON blob it would probably not be a pure function of cluster id.

So from my current understanding of the GCE API:

+1 to JSON being the answer if we want to add metadata in the future, but -1 to JSON being the answer if we need to do fancier filtering in the future, so I'd rather leave it as is for now and let future feature decisions decide which way to go.

Absolutely if we need more metadata in the future we should do that though. Shame on the GCE API about not having better ways to filter.

sarum90 commented:
Got all greens (Including the GCE functional tests) So I'll be merging now.
